specs/core/1_shard-data-chains.md

```diff
@@ -17,596 +17,384 @@
+    active_validator_indices = get_active_validator_indices(beacon_state, epoch)
+    seed = get_seed(beacon_state, epoch, DOMAIN_SHARD_ATTESTER)
+    return compute_committee(active_validator_indices, seed, shard, SHARD_COUNT)[:MAX_PERIOD_COMMITTEE_SIZE]
 ```
 
-### `get_persistent_committee`
+#### `get_shard_committee`
 
 ```python
-def get_persistent_committee(state: BeaconState,
-                             shard: Shard,
-                             slot: ShardSlot) -> Sequence[ValidatorIndex]:
-    """
-    Return the persistent committee for the given ``shard`` at the given ``slot``.
-    """
-    epoch = compute_epoch_of_shard_slot(slot)
-
-    earlier_committee = get_period_committee(state, get_shard_period_start_epoch(epoch, lookback=2), shard)
-    later_committee = get_period_committee(state, get_shard_period_start_epoch(epoch, lookback=1), shard)
-
-    # Take not-yet-cycled-out validators from earlier committee and already-cycled-in validators from
-    # later committee; return a sorted list of the union of the two, deduplicated
-    return sorted(set(
-        [i for i in earlier_committee if epoch % EPOCHS_PER_SHARD_PERIOD < i % EPOCHS_PER_SHARD_PERIOD]
-        + [i for i in later_committee if epoch % EPOCHS_PER_SHARD_PERIOD >= i % EPOCHS_PER_SHARD_PERIOD]
-    ))
+def get_shard_committee(beacon_state: BeaconState, shard: Shard, epoch: Epoch) -> Sequence[ValidatorIndex]:
+    older_committee = get_period_committee(beacon_state, shard, compute_shard_period_start_epoch(epoch, 2))
+    newer_committee = get_period_committee(beacon_state, shard, compute_shard_period_start_epoch(epoch, 1))
+    # Every epoch cycle out validators from the older committee and cycle in validators from the newer committee
+    older_subcommittee = [i for i in older_committee if i % EPOCHS_PER_SHARD_PERIOD > epoch % EPOCHS_PER_SHARD_PERIOD]
+    newer_subcommittee = [i for i in newer_committee if i % EPOCHS_PER_SHARD_PERIOD <= epoch % EPOCHS_PER_SHARD_PERIOD]
+    return older_subcommittee + newer_subcommittee
 ```
 
-### `get_shard_block_proposer_index`
+#### `get_shard_proposer_index`
 
 ```python
-def get_shard_block_proposer_index(state: BeaconState,
-                                   shard: Shard,
-                                   slot: ShardSlot) -> Optional[ValidatorIndex]:
-    # Randomly shift persistent committee
-    persistent_committee = list(get_persistent_committee(state, shard, slot))
-    current_epoch = get_current_epoch(state)
-
-    active_indices = [i for i in persistent_committee if is_active_validator(state.validators[i], current_epoch)]
+def get_shard_proposer_index(beacon_state: BeaconState, shard: Shard, slot: ShardSlot) -> Optional[ValidatorIndex]:
+    epoch = get_current_epoch(beacon_state)
+    shard_committee = get_shard_committee(beacon_state, shard, epoch)
+    active_indices = [i for i in shard_committee if is_active_validator(beacon_state.validators[i], epoch)]
     if not any(active_indices):
```

Comment:
handled merge @JustinDrake.

I added this `if not any` clause back in in the spirit of being able to have extremely low validator count. `shard_committee` length is not guaranteed to be non-zero and if it is empty, `compute_proposer_index` will never terminate.